source/val/validate_memory.cpp

```diff
@@ -1181,6 +1181,21 @@ spv_result_t ValidateArrayLength(ValidationState_t& state,
            << state.getIdName(inst->id()) << "' must be an OpTypeRuntimeArray.";
   }
 
+  const bool uses_variable_pointers =
+      state.features().variable_pointers ||
+      state.features().variable_pointers_storage_buffer;
+
+  if (!pointer_type ||
```

Comment:
This doesn't capture the spec correctly. If variable pointers are not enabled you will accept OpAccessChain as the pointer operand, but not if they are enabled.

The check needs to be more subtle. https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#VariablePointer defines the roots and transitivity of variable pointers. You will have to trace the generation of the pointer operand to determine if it is a variable pointer or not.